### ISSUE TITLE
Enable sharing of browserlist among projects

### ISSUE THREAD

**Issue by morewry** [ai/browserslist#122]
This project addresses the need to share a browerlist among tools, but I think it's equally important to be able to share a list among multiple modules. Tools such as stylelint and eslint enable sharing configurations with `extends` properties that correlate to the name of a preset installed as a Node.js. 

The biggest reason this is useful is that you would typically want to base your supported browsers on the traffic to _your_ website (rather than aggregate traffic to other websites). While you can generate a browserlist from your website stats and use it, it currently looks like you would have to make and update multiple copies of the list. Unless this is something enabled via babel-env-preset or another option. (It's not apparent to me, but I might be missing it).

Adding a feature that enables a user to have and use a browser list module would make this configuration even more useful than it is already.

e.g. in a `package.json`

```json
...
"browserlist": {
  "extends": "my-browserlist"
}
...
```

```shell
npm install --save-dev browserlist-preset-my-browserlist
```

**Comment by morewry** [comment]
It looks like,

> You can specify direct path in BROWSERSLIST_CONFIG environment variables.

This could serve a similar purpose. It's preferable to avoid having to maintain specific paths, but it addresses the general need.

**Comment by morewry** [comment]
> Save the data to a browserslist-stats.json file in your project.

If this has to be saved in every project, I'm back to needing a way to share it. :-)

**Comment by ai** [comment]
Hm. Yeap, I like Stylelint/ESLint `extends` too. But I think in Browserslist case `BROWSERSLIST_CONFIG` will be enough.

There are few difference with Stylelint/ESLint:

* Browserslist config is smaller. In most cases it is just few lines.
* In linters you must to define your rules. In Browserslist it is better to have no `browserslist` config — defaults is better for most of project. I afraid that mobile web developers will start to use `last 2 iOS versions` and we will be back in `-webkit-` hell :(.
* I hope `browserslist` config will be used by non-JS tools. But `extends` force to use `npm`.

But there is no really important reason in that reasons. So let’s keep it open to get feedback, maybe people really want `extends` and I am wrong :D.

**Comment by ai** [comment]
We have no feedback, so I will close issue

**Comment by destroyerofbuilds** [comment]
I just came across this issue, and wanted to share my own setup in hopes it provides context on why supporting an `extends` like feature may be valuable.

At the company I work for our customers use a desktop application that ships with an embedded version of Chrome. Customers may be using any version of our desktop application released within the past few years, so they may be using one of several versions of Chrome.

We ship quite a few applications to our consumers through the embedded browser. Because a consumer can access all applications from any version of the desktop application, we follow the practice of always transpiling to the oldest version of Chrome used in a supported version of our desktop application.

For Babel we wrote our own preset on top of `babel-preset-env`. Our preset actually supports targets, allowing developers to say whether their project will run as an application within our desktop application, or is a headless service running on Node.

For a Node service you may see something like the following within one of our projects:

```json
cat .babelrc
{
  "presets": [
    ["babel-preset-ourcompany", {
      "targets": ["node"]
    }]
  ]
}
```

If the project is an application destined to be shipped through our desktop application, they may add `desktop-application` to `target` list, causing our preset to pass the appropriate Chrome versions to `babel-preset-env`. 

When we retire a version of our desktop application, which means we remove a version of Chrome we need to support, we release a new major version of our preset, bumping the minimum Chrome browser used by Babel.

At this point the major version slowly rolls out to all our projects. 

By packaging the browser list via a Babel preset we're able to coordinate our browser compatibility across all applications. 

For handling CSS, we use the `postcss-loader` with the `autoprefixer` plugin.

Furthermore we're using those with Webpack.

So here is where we run into a slightly less optimal approach to handling browser compatibility.

While we are able to ship a Babel preset that allows all projects to transpile JavaScript in a manner that makes their application compatible with all our supported versions of Chrome, we can't do the same with `autoprefixer`. 

So we end up with the following in our shared Webpack configuration:

```javascript
{
  loader: require.resolve('postcss-loader'),
  options: {
    plugins: () => {
      return [
        autoprefixer({
          // TODO: This browser list needs to be refactored out so that we can share the list
          // with our `babel-preset-ourcompany` package. `babel-preset-ourcompany` allows
          // users to specify a target platform(s). Based on that list, we
          // generate a list of browsers that `babel` should support during transpilation.
          // * How do we build a browser list here that allows users to target a platform?
          // * How do we share a target platform list with `babel-preset-ourcompany`?
          // * Perhaps use shared `browserlist` - https://github.com/ai/browserslist
          browsers: [
            // browser list
          ],
        }),
      ];
    },
  },
},
```

I left a few of our comments in to provide context on what we were thinking when we wrote our setup.

**Comment by ai** [comment]
@destroyerofbuilds what about:

```js
autoprefixer({ browsers: require("@company/browsers") })
```

This npm module will just return a array with browsers?

But I agree, that we have some question for Browserslist.

**Comment by destroyerofbuilds** [comment]
@ai for the time being that's probably a good enough solution to help us consolidate our duplicated lists.

We'd probably also look to `require` the same package within our `babel-preset-ourcompany` package as well. 

Why I said _good enough_ is that I'd like to, at some point, leverage the `.browserslistrc` file so that we don't need to maintain the extra layer of indirection (the `require` within a package underneath the application itself), but could just have teams directly specify the browser lists they need to support, and then tools like Babel or Autoprefixer, can just use the `.browserslistrc` file directly. 

So, thank you very much for the idea, and if I remember, I'll provide feedback on how that approach works out.

**Comment by destroyerofbuilds** [comment]
Also, for a little clarification, one of my comments above, in the quoted webpack configuration, mentions multiple target platforms.

We don't just ship applications through a single desktop application. We actually have several distribution platforms. So some projects may provide `desktop-application` as a target in Babel, while others may include that target and `other-distribution-platform`, and within our `babel-preset-ourcompany` preset, we append two browser lists together (roughly speaking) so that the application will be transpiled to work on browsers used with either distribution platform.

**Comment by ai** [comment]
What do you think about this syntax:

```json
  "browserslist": {
    "use company",
    "last 2 versions"
  }
```

And it will load `browserslist-config-company`

**Comment by destroyerofbuilds** [comment]
@ai does that syntax establish the large assumption that the consumer is using Node modules? You mentioned earlier: `I hope browserslist config will be used by non-JS tools.`

Would `use ./PATH/TO/browserllist.file` work?

Setting up a scenario:

An application team `npm install`'s `browserlist-config-company`.

They then add a `.browserlistrc` file at the root of their project (or in their `package.json` file)

```
use ./node_modules/browserlist-config-company/desktop-application
last 2 versions
```

If they deploy to both of our distribution platforms:

```
use ./node_modules/browserlist-config-company/desktop-application
use ./node_modules/browserlist-config-company/other-distribution-platform
last 2 versions
```

**Comment by ai** [comment]
@destroyerofbuilds good question about be independent from npm. Maybe `use company` will look into `npm`, but if somebody we have real case we will add `use file:path/to/config`.

**Comment by destroyerofbuilds** [comment]
@ai so assuming we can leverage the node module resolution approach to load shared browser configuration, what should the file format be?

Also, how would we handle the situation of some applications needing to load two browser lists? Will we need to publish two browserlist packages?

**Comment by ai** [comment]
@destroyerofbuilds it will be CommonJS module with array of strings.

```js
module.exports = ["last 1 version"]
```

You can write `name/path` to load `require("browserslist-config-name/path")`

**Comment by destroyerofbuilds** [comment]
Sounds great! :smile:

**Comment by ai** [comment]
Done https://github.com/ai/browserslist/pull/171

**Comment by ai** [comment]
Released in 2.5
